pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -497,36 +496,28 @@ func getEvenPodsSpreadMetadata(pod *v1.Pod, allNodes []*schedulernodeinfo.NodeIn
 	return &m, nil
 }
 
-func getHardTopologySpreadConstraints(pod *v1.Pod) (constraints []v1.TopologySpreadConstraint) {
-	if pod != nil {
-		for _, constraint := range pod.Spec.TopologySpreadConstraints {
-			if constraint.WhenUnsatisfiable == v1.DoNotSchedule {
-				constraints = append(constraints, constraint)
+func filterHardTopologySpreadConstraints(constraints []v1.TopologySpreadConstraint) ([]topologySpreadConstraint, error) {
+	var result []topologySpreadConstraint
+	for _, c := range constraints {
+		if c.WhenUnsatisfiable == v1.DoNotSchedule {
+			selector, err := metav1.LabelSelectorAsSelector(c.LabelSelector)
```

Comment:
It seems this is the key perf of the improvement - `LabelSelectorAsSelector` doesn't sound like a cheap operation, and if we cache the result, we seem to get a lot of benefits.

A followup is to do a thorough check of "repeated usage of LabelSelectorAsSelector" and do similar refactoring. I do recall there was a PR using the same rationale improving the performance of PodAffinity, probably #79465. We should consolidate this.

Reply:
sounds like suggestion. cc @liu-cong who is looking into performance in general.

Reply:
We should run some benchmarks first, but we could annotate the pods in the cache with already processed Selectors for affinity and spread constraints, and so we do this once per across scheduling cycles, not just once per cycle.

Reply:
SG:) I was discussing with ahg-g yesterday about what can be cached per pod and this seems to fall into that category.